torch/testing/_internal/common_methods_invocations.py

```diff
@@ -2730,6 +2784,34 @@ def reference_sgn(x):
                    dtypesIfCPU=all_types_and_complex_and(torch.half, torch.bfloat16),
                    dtypesIfCUDA=all_types_and_complex_and(torch.half, torch.bfloat16),
                    assert_autodiffed=True,),
+    OpInfo('pow',
+           dtypes=all_types_and_complex_and(torch.half, torch.bfloat16),
+           sample_inputs_func=sample_inputs_pow,
+           supports_inplace_autograd=False,
+           assert_autodiffed=True,
+           skips=(
+               # Inplace variants currently use the same sample inputs as the method for which
+               # they are a variant, causing this test to fail because of broadcasting semantics.
+               # This test should be enabled after GitHub PR 53014 would land.
+               # Reference: https://github.com/pytorch/pytorch/issues/50747
+               SkipInfo('TestCommon', 'test_variant_consistency_eager',
+                        dtypes=[torch.float32, torch.float64]),
+               # Due to AVX2 curently not being fully supported for Float16, log_vml_cpu can't be enabled
+               # for Float16, causing this test to fail.
+               SkipInfo('TestOpInfo', 'test_supported_backward',
+                        device_type='cpu', dtypes=[torch.float16]),
+               # JIT doesn't currently support complex literals.
+               # This test should be enabled after GitHub PR 52881 would land.
+               # Reference: https://github.com/pytorch/pytorch/issues/51996#issuecomment-791517296
+               SkipInfo('TestCommon', 'test_variant_consistency_jit',
+                        dtypes=[torch.complex64, torch.complex128]),
+               # Bool is not present in the dtypes list, but pow on bool tensors is currently supported in the
+               # sense that they're casted to other dtypes, so there's no point in running OpInfo tests on bool.
+               # Moreover, adding bool to the dtypes list results in Runtime errors when tests are run for it.
+               # But since invoking torch.pow() on a tensor with dtype bool doesn't result in a Runtime error,
+               # as otherwise expected by this test, it needs to be skipped.
+               SkipInfo('TestOpInfo', 'test_unsupported_dtypes', dtypes=[torch.bool]),)
```

Comment:
Is this because the inplace version of pow (`pow_`) is not supported for `bool` as it tries to write an int result into the bool self tensor?

Reply:
The runtime error was being thrown by pybind11 but you might be right. I'll look it up.

Reply:
@heitorschueroff, should OpInfo tests even be run for bool, given that bool tensors are cast to either long, float, or double tensors?

Reply:
If I add `torch.bool` to supported dtypes, then `test_supported_dtypes_pow_cpu_bool` and `test_supported_dtypes_pow_cuda_bool` fail with errors `RuntimeError: "pow" not implemented for 'Bool'` and `RuntimeError: "pow_cuda" not implemented for 'Bool'` respectively.

That's why I haven't added `torch.bool` to the `dtypes` list. 
As of now, `test_unsupported_dtypes` has to be skipped for it because it is supported in the sense that the result is cast to another dtype, so it fails with `AssertionError: RuntimeError not raised`.

Reply:
Yes, casting bool to double is consistent with `np.power`. However, this introduces a problem with the inplace variant which tries to write the `double` result back into the Tensor with dtype `bool`. I think skipping it is ok for now but could you add a comment explaining that the problem is with the inplace variant.

Reply:
@heitorschueroff, thank you for your response!

However, I think that the issue is not being caused by the inplace variant.

`test_supported_dtypes` does not test the inplace-variant but pybind11 still throws an exception that becomes a RuntimeError, `RuntimeError: "pow" not implemented for 'Bool'`: 
https://github.com/pytorch/pytorch/blob/master/test/test_ops.py#L53-L55

Reply:
@heitorschueroff 

`test_supported_dtypes` for `torch.bool` is called for the _function variant_ of `pow`, and fails due to an exception thrown in `PowKernel.cpp` (by `pow_tensor_tensor_kernel`), as `iter.common_dtype()` is `ScalarType::Bool`.
So, the _function variant_ is unsupported for `bool`. However, it is supported by numpy.

Only its _method variant_ is supported by PyTorch. So, I'll modify the comments accordingly.
But does the _function variant_ of `pow` need to be supported for `bool`, in order to be consistent with `numpy`?

Reply:
@imaginary-person Great work tracking down the source of this error. We should support `bool` for the function variant as well but that can be done in another PR. Would you be so kind as to open an issue for this?

FYI: The inplace variant might not be the one causing this error but it does cause a error
```python
>>> torch.tensor(2, dtype=torch.bool).pow_(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Found dtype Bool but expected Long
```

Reply:
Sure, I'll open an issue for it asap.

BTW, does the inplace variant for `pow` need to be supported for `bool` as well?

Reply:
> Yes, casting bool to double is consistent with `np.power`. However, this introduces a problem with the inplace variant which tries to write the `double` result back into the Tensor with dtype `bool`. I think skipping it is ok for now but could you add a comment explaining that the problem is with the inplace variant.

Remember that these operations cast to the default floating point type. In NumPy the default floating point type is double. In PyTorch the default floating point type is float32 (although we allow the user to set it).

Reply:
Hello @mruberry,

> Why does `pow(bool, bool)` promote to floating point, however?
It doesn't. It throws a runtime error simply because it's not implemented in PowKernel.cpp.

> pow supporting bool inputs seems very niche; if we didn't support it previously I wouldn't bother including support for it now unless it's very easy to do.

I completely agree. I can't think of any use-case. The only issue we were facing is related to running tests for it.
I haven't included `bool` in the dtypes list in its `OpInfo`, BTW.

As for the implementation, all values in the output tensor would be 1.
So, if `pow(bool tensor, bool tensor)` needs to be supported, a conditional statement can be added to generate such a tensor for the `(bool, bool)` case in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Pow.cpp#L12

Reply:
Oh I see. When a bool tensor is combined with a double scalar it promotes to the default floating point type (float32). That makes sense and that does work in current PyTorch. It also doesn't work inplace in the current PyTorch, and that's OK to keep it that way. The supported dtypes test for bool should work if the SampleInputs for the bool dtype are constructed properly, though? That is, I would expect the SampleInput for the bool dtype to be a bool tensor and a floating point Python scalar, for example.

Reply:
Thanks for your response, @mruberry! I'll add bool to the `dtypes` list & add some sample inputs specifically for it.

Reply:
Let me or @heitorschueroff know if you have any questions, @imaginary-person. Thanks for paying such close attention to this issue.